pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java

```diff
@@ -245,6 +245,13 @@
     )
     private int numIOThreads = 2 * Runtime.getRuntime().availableProcessors();
 
+    @FieldContext(
+            category = CATEGORY_SERVER,
+            doc = "Number of shared threads to use for broker clients."
+                    + " Default is set to `2 * Runtime.getRuntime().availableProcessors()`"
+    )
+    private int brokerClientNumIOThreads = 2 * Runtime.getRuntime().availableProcessors();
```

Comment:
I'd say that we should prob reuse the broker IO threads here

Reply:
The problem is that broker IO threads is a `EventLoopGroup` instance which is already used in the Pulsar Client.

https://github.com/apache/pulsar/blob/9032a9afc15e029d8f205761b5e104f44f9c4be0/pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java#L319-L320

The Pulsar Client has this internal/IO and external/listener executors in addition to the `eventLoopGroup`.

https://github.com/apache/pulsar/blob/4591a210a8d924d99811edfdfeb7433452e5eeb4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L182-L185

It might be a risky change in Pulsar Client to replace the internal executor to delegate directly to `eventLoopGroup`.

This is the reason why I think it's necessary to have a separate setting to control the amount of threads used to run the Pulsar Client internal executors. Before this PR each client has had a single threaded internal executor for each client instance.

Reply:
Yes, I think the confusion comes from the name `brokerClientNumIOThreads`, since it's client internal executor and not the client event loop which is already shared. 

Additionally, the client internal executor shouldn't really be used a lot in broker, since it's used mainly for dispatching the partitioned/multi topic consumers, so we could just default to 1 thread instead.

Reply:
> Yes, I think the confusion comes from the name brokerClientNumIOThreads, since it's client internal executor and not the client event loop which is already shared.

That's true. The Pulsar Client uses the `numIOThreads` parameter for configuring the number of threads for the `eventLoopGroup` and also for the internal/IO executor.

https://github.com/apache/pulsar/blob/4591a210a8d924d99811edfdfeb7433452e5eeb4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L979

https://github.com/apache/pulsar/blob/4591a210a8d924d99811edfdfeb7433452e5eeb4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java#L184-L185

This is confusing. The internal executor was introduced in #8208 and that change uses `numIOThreads` parameter also for configuring the number of threads for the internal executor. @merlimat I wonder what would be a way to improve this situation?

Reply:
> Additionally, the client internal executor shouldn't really be used a lot in broker, since it's used mainly for dispatching the partitioned/multi topic consumers, so we could just default to 1 thread instead.

yes, that is true. I wonder if the value needs to be configurable at all? I'll remove brokerClientNumIOThreads key and just use the value 1 as I did for the external executor.

Reply:
I made the changes now. @merlimat PTAL

Reply:
👍